flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java

```diff
@@ -965,11 +965,20 @@ public void cancel() {
 
 					// we build a future that is complete once all vertices have reached a terminal state
 					final ConjunctFuture<Void> allTerminal = FutureUtils.waitForAll(futures);
-					allTerminal.thenAccept(
-						(Void value) -> {
-							// cancellations may currently be overridden by failures which trigger
-							// restarts, so we need to pass a proper restart global version here
-							allVerticesInTerminalState(globalVersionForRestart);
+					allTerminal.whenCompleteAsync(
```

Comment:
Does it have to run asynchronously? If yes, does it make sense to specify a thread pool? Now it can run on `ForkJoinPool.commonPool()`.

Reply:
It does not have to run asynchronously. I will remove it.

We need the `transitionState` because it might happen that we cannot properly cancel all `ExecutionJobVertices`. E.g. if the `SlotOwner` cannot return the allocated slot. This is currently not the case, but we should guard against it if this should change in the future.